### PR TITLE
vic-machine suggests datacenter

### LINKED ISSUE #1711: vic-machine: suggest valid datastore, networks and datacenter
Similar to compute resource suggestion, we also need suggestion for datacenter, network and datastore

### PR DESCRIPTION
**Pull request by andrewtchin**
Fixes #1711 

RP specified in target:
```
⇒  bin/vic-machine-linux create --target=10.17.109.115/dc1/asdf --user=Administrator@vsphere.local --image-store=vsanDatastore --password=password --force=true  --compute-resource=cluster1/chin --name=VCH-test
INFO[2016-08-12T09:26:08-07:00] ### Installing VCH ####                      
INFO[2016-08-12T09:26:08-07:00] Generating certificate/key pair - private key in ./VCH-test-key.pem 
ERRO[2016-08-12T09:26:09-07:00] --target should only specify datacenter in the path (e.g. https://addr/datacenter) - specify cluster, resource pool, or folder with --compute-resource 
INFO[2016-08-12T09:26:09-07:00] Suggesting valid values for datacenter in --target 
INFO[2016-08-12T09:26:09-07:00] Suggested datacenters:                       
INFO[2016-08-12T09:26:09-07:00]   "dc1"                                      
INFO[2016-08-12T09:26:09-07:00]   "dc2"                                      
ERRO[2016-08-12T09:26:09-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:26:09-07:00] --------------------                         
ERRO[2016-08-12T09:26:09-07:00] vic-machine-linux failed: --target should only specify datacenter in the path (e.g. https://addr/datacenter) - specify cluster, resource pool, or folder with --compute-resource
 ```

Datacenter DNE:
```
⇒  bin/vic-machine-linux create --target=10.17.109.115/asdf --user=Administrator@vsphere.local --image-store=vsanDatastore --password=password --force=true --compute-resource=cluster1/chin - --name=VCH-test  
INFO[2016-08-12T09:26:23-07:00] ### Installing VCH ####                      
INFO[2016-08-12T09:26:23-07:00] Generating certificate/key pair - private key in ./VCH-test-key.pem 
ERRO[2016-08-12T09:26:24-07:00] Datacenter must be specified in --target (e.g. https://addr/datacenter) 
INFO[2016-08-12T09:26:24-07:00] Suggesting valid values for datacenter in --target 
INFO[2016-08-12T09:26:24-07:00] Suggested datacenters:                       
INFO[2016-08-12T09:26:24-07:00]   "dc1"                                      
INFO[2016-08-12T09:26:24-07:00]   "dc2"                                      
ERRO[2016-08-12T09:26:24-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:26:24-07:00] --------------------                         
ERRO[2016-08-12T09:26:24-07:00] vic-machine-linux failed: Datacenter must be specified in --target (e.g. https://addr/datacenter)
```

Datacenter not specified when multiple exist in vC:
```
⇒  bin/vic-machine-linux create --target=10.17.109.115 --user=Administrator@vsphere.local --image-store=vsanDatastore --password=password --force=true --compute-resource=cluster1/chin --name=VCH-test
INFO[2016-08-12T09:34:47-07:00] ### Installing VCH ####                      
INFO[2016-08-12T09:34:47-07:00] Generating certificate/key pair - private key in ./VCH-test-key.pem 
ERRO[2016-08-12T09:34:48-07:00] Datacenter must be specified in --target (e.g. https://addr/datacenter) 
INFO[2016-08-12T09:34:48-07:00] Suggesting valid values for datacenter in --target 
INFO[2016-08-12T09:34:48-07:00] Suggested datacenters:                       
INFO[2016-08-12T09:34:48-07:00]   "dc1"                                      
INFO[2016-08-12T09:34:48-07:00]   "dc2"                                      
ERRO[2016-08-12T09:34:48-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:34:48-07:00] --------------------                         
ERRO[2016-08-12T09:34:48-07:00] vic-machine-linux failed: Datacenter must be specified in --target (e.g. https://addr/datacenter)
```

ESX
```
⇒  bin/vic-machine-linux create --target 192.168.218.177/asdf --name chin-vic-110 --user root --password password --image-store datastore1                                                                                                                                                                   
INFO[2016-08-12T09:39:56-07:00] ### Installing VCH ####                      
INFO[2016-08-12T09:39:56-07:00] Generating certificate/key pair - private key in ./chin-vic-110-key.pem 
ERRO[2016-08-12T09:39:56-07:00] Datacenter must be specified in --target (e.g. https://addr/datacenter) 
INFO[2016-08-12T09:39:56-07:00] Suggesting valid values for datacenter in --target 
INFO[2016-08-12T09:39:56-07:00] Suggested datacenters:                       
INFO[2016-08-12T09:39:56-07:00]   "ha-datacenter"                            
ERRO[2016-08-12T09:39:56-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:39:56-07:00] --------------------                         
ERRO[2016-08-12T09:39:56-07:00] vic-machine-linux failed: Datacenter must be specified in --target (e.g. https://addr/datacenter)
 
⇒  bin/vic-machine-linux create --target 192.168.218.177/asdf/asdf --name chin-vic-110 --user root --password password --image-store datastore1
INFO[2016-08-12T09:40:07-07:00] ### Installing VCH ####                      
INFO[2016-08-12T09:40:07-07:00] Generating certificate/key pair - private key in ./chin-vic-110-key.pem 
ERRO[2016-08-12T09:40:07-07:00] --target should only specify datacenter in the path (e.g. https://addr/datacenter) - specify cluster, resource pool, or folder with --compute-resource 
INFO[2016-08-12T09:40:07-07:00] Suggesting valid values for datacenter in --target 
INFO[2016-08-12T09:40:07-07:00] Suggested datacenters:                       
INFO[2016-08-12T09:40:07-07:00]   "ha-datacenter"                            
ERRO[2016-08-12T09:40:07-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:40:07-07:00] --------------------                         
ERRO[2016-08-12T09:40:07-07:00] vic-machine-linux failed: --target should only specify datacenter in the path (e.g. https://addr/datacenter) - specify cluster, resource pool, or folder with --compute-resource
```

local regression tests pass

### CODE REVIEW

File: lib/install/validate/validator.go [hunk 1]
```diff
@@ -127,20 +118,60 @@ func CreateNoDCCheck(ctx context.Context, input *data.Data) (*Validator, error)
...
+	log.Info("Suggesting valid values for datacenter in --target")
+
+	dcs, err := v.Session.Finder.DatacenterList(v.Context, "*")
+	if err != nil {
+		log.Errorf("Unable to list datacenters: %s", err)
+		return
+	}
+
+	if len(dcs) == 0 {
+		log.Info("No datacenters found")
```
Comment (gigawhitlocks) [inline lib/install/validate/validator.go]: `log.Errorf` instead here?

File: lib/install/validate/validator.go [hunk 2]
```diff
@@ -127,20 +118,60 @@ func CreateNoDCCheck(ctx context.Context, input *data.Data) (*Validator, error)
...
+		return nil, errors.New(detail)
+	}
+
 	return v, nil
 }
 
 func (v *Validator) datacenter() error {
 	if v.Session.Datacenter == nil {
-		detail := "Target should specify datacenter when there are multiple possibilities, e.g. https://addr/datacenter"
+		detail := "Datacenter must be specified in --target (e.g. https://addr/datacenter)"
```
Comment (gigawhitlocks) [inline lib/install/validate/validator.go]: datacenter url starts with `https`?

**Comment by gigawhitlocks** [comment]
Can we get an integration test to protect this feature added to the vic machine tests?

**Comment by andrewtchin** [comment]
It can be done

File: lib/install/validate/validator.go (hunk 2 above)
Comment (andrewtchin) [inline lib/install/validate/validator.go]: yes https://github.com/vmware/vic/blob/master/lib/install/validate/validator.go#L85

File: lib/install/validate/validator.go (hunk 1 above)
Comment (andrewtchin) [inline lib/install/validate/validator.go]: I'm going to leave this as it's consistent with the other suggestions and we didn't fail to query the environment just got no results

File: lib/install/validate/validator.go (hunk 2 above)
Comment (gigawhitlocks) [inline lib/install/validate/validator.go]: HUH! TIL thanks 👍

File: lib/install/validate/validator.go (hunk 1 above)
Comment (gigawhitlocks) [inline lib/install/validate/validator.go]: Not finding any datacenters seems fatal to me though, am I wrong? At this point the user-supplied search failed as did our search for _anything_ -- definitely go with your intuition here as you definitely have more understanding & context than I do, but I assume that on this code-path installation does abort during validation?

Comment (andrewtchin) [inline lib/install/validate/validator.go]: If we get to suggesting datacenters we are already going to abort the install with a message like below from either line 124 or 135
```
ERRO[2016-08-12T09:40:07-07:00] Create cannot continue: failed to create validator 
ERRO[2016-08-12T09:40:07-07:00] --------------------                         
ERRO[2016-08-12T09:40:07-07:00] vic-machine-linux failed: --target should only specify datacenter in the path (e.g. https://addr/datacenter) - specify cluster, resource pool, or folder with --compute-resource
```

File: lib/install/validate/validator.go [hunk 3]
```diff
@@ -127,20 +118,60 @@ func CreateNoDCCheck(ctx context.Context, input *data.Data) (*Validator, error)
 
 	v.Session.Populate(ctx)
 
+	// only allow the datacenter to be specified in the taget url, if any
```
Comment (emlin) [inline lib/install/validate/validator.go]: session.populate will search datacenter, if don't set before populate, there will have datacenter not set error.

Comment (emlin) [inline lib/install/validate/validator.go]: never mind, you just moved the error message.

**Comment by emlin** [comment]
lgtm

**Comment by andrewtchin** [comment]
self lgtm

**Comment by andrewtchin** [comment]
♻️

**Pull request merged**